install/0000_90_machine-config-operator_01_prometheus-rules.yaml

```diff
@@ -103,3 +103,67 @@ spec:
+            > 60
+          for: 10m
+          labels:
+            namespace: openshift-kube-apiserver
+            severity: warning
+        - alert: ExtremelyHighIndividualControlPlaneCPU
+          annotations:
+            summary: >-
+              CPU utilization on a single control plane node is very high, more CPU pressure is likely to cause a failover; increase available CPU.
+            runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/machine-config-operator/ExtremelyHighIndividualControlPlaneCPU.md
+            description: >-
+              Extreme CPU pressure can cause slow serialization and poor performance from the kube-apiserver and etcd.
+              When this happens, there is a risk of clients seeing non-responsive API requests which are issued again
+              causing even more CPU pressure.
+              It can also cause failing liveness probes due to slow etcd responsiveness on the backend.
+              If one kube-apiserver fails under this condition, chances are you will experience a cascade as the remaining
+              kube-apiservers are also under-provisioned.
+              To fix this, increase the CPU and memory on your control plane nodes.
+          expr: |
+            100 - (avg by (instance) (rate(node_cpu_seconds_total{mode="idle"}[1m])) * 100) > 90 AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+          for: 5m
+          labels:
+            namespace: openshift-kube-apiserver
+            severity: warning
+        - alert: ExtremelyHighIndividualControlPlaneCPU
+          annotations:
+            summary: >-
+              Sustained high CPU utilization on a single control plane node, more CPU pressure is likely to cause a failover; increase available CPU.
+            runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/machine-config-operator/ExtremelyHighIndividualControlPlaneCPU.md
+            description: >-
+              Extreme CPU pressure can cause slow serialization and poor performance from the kube-apiserver and etcd.
+              When this happens, there is a risk of clients seeing non-responsive API requests which are issued again
+              causing even more CPU pressure.
+              It can also cause failing liveness probes due to slow etcd responsiveness on the backend.
+              If one kube-apiserver fails under this condition, chances are you will experience a cascade as the remaining
+              kube-apiservers are also under-provisioned.
+              To fix this, increase the CPU and memory on your control plane nodes.
+          expr: |
+            100 - (avg by (instance) (rate(node_cpu_seconds_total{mode="idle"}[1m])) * 100) > 90 AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+          for: 1h
+          labels:
+            namespace: openshift-kube-apiserver
+            severity: critical
```

Comment:
When this `critical` alert goes off, can we set up some kind of inhibitor to mask the `warning`-level 5m version of this alert?  Should those two versions get different names?  If they're worth different `summary` and `description`, I would expect them to be worth different names.

Reply:
Yes, I think having an inhibitor makes a lot of sense here. I don't if we have any precedence for that in OCP, but we could improve a lot of alerts like that.

> Should those two versions get different names? If they're worth different summary and description, I would expect them to be worth different names.

There is barely any difference in the summary and description of these alerts. These alerts are basically the same but depending on the disruption time, the criticality of the alert will change. That being said, I don't think the current approach makes sense. I wouldn't expect any customer to take a look at a warning alert in less than an hour, which renders the warning level alert useless in practice. Rather than having the criticality dependent on the disruption time, I would rather have it depending on the CPU usage. For example, we could change the warning level alert to be 70-80% CPU usage over 1h. wdyt?

Reply:
> Rather than having the criticality dependent on the disruption time, I would rather have it depending on the CPU usage. For example, we could change the warning level alert to be 70-80% CPU usage over 1h.

Yeah, sounds good.  But use an inhibitor, because you want `warning` firing if the node is oscillating between 75 and 95, but not staying >90 long enough to set off the `critical`.